File: implementations/java/provekit-ir/src/main/java/com/provekit/ir/Declaration.java
package com.provekit.ir;

/**
 * IR Document declarations matching the CDDL spec.
 */
public sealed interface Declaration {
    String toJson();

    record Property(String name, Param[] params, Formula body) implements Declaration {
        public String toJson() {
            StringBuilder sb = new StringBuilder("{\"kind\":\"property\",\"name\":\"" + Sort.escape(name) + "\",\"params\":[");
            for (int i = 0; i < params.length; i++) {
                if (i > 0) sb.append(",");
                sb.append(params[i].toJson());
            }
            sb.append("],\"body\":").append(body.toJson()).append("}");
            return sb.toString();
        }
    }

    record Param(String name, Sort sort) {
        public String toJson() {
            return "{\"name\":\"" + Sort.escape(name) + "\",\"sort\":" + sort.toJson() + "}";
        }
    }

    /**
     * BridgeDeclaration per IR formal grammar v1.1.0
     * (protocol/specs/2026-04-30-ir-formal-grammar.md).
     *
     * Required fields: name, sourceSymbol, sourceLayer, sourceContractCid,
     * targetContractCid, targetProofCid, targetLayer.
     * Optional: notes (omitted from output when null).
     *
     * JCS canonical key order (RFC 8785, alphabetical by code unit):
     * kind, name, [notes,] sourceContractCid, sourceLayer, sourceSymbol,
     * targetContractCid, targetLayer, targetProofCid.
     */
    record Bridge(
        String name,
        String sourceSymbol,
        String sourceLayer,
        String sourceContractCid,
        String targetContractCid,
        String targetProofCid,
        String targetLayer,
        String notes
    ) implements Declaration {
        public String toJson() {
            StringBuilder sb = new StringBuilder();
            sb.append("{\"kind\":\"bridge\"");
            sb.append(",\"name\":\"").append(Sort.escape(name)).append("\"");
            if (notes != null) {
                sb.append(",\"notes\":\"").append(Sort.escape(notes)).append("\"");
            }
            sb.append(",\"sourceContractCid\":\"").append(Sort.escape(sourceContractCid)).append("\"");
            sb.append(",\"sourceLayer\":\"").append(Sort.escape(sourceLayer)).append("\"");
            sb.append(",\"sourceSymbol\":\"").append(Sort.escape(sourceSymbol)).append("\"");
            sb.append(",\"targetContractCid\":\"").append(Sort.escape(targetContractCid)).append("\"");
            sb.append(",\"targetLayer\":\"").append(Sort.escape(targetLayer)).append("\"");
            sb.append(",\"targetProofCid\":\"").append(Sort.escape(targetProofCid)).append("\"");
            sb.append("}");
            return sb.toString();
        }
    }

    record Contract(String name, String outBinding, Formula pre, Formula post, Formula inv, String evidence) implements Declaration {
        public String toJson() {
            StringBuilder sb = new StringBuilder("{\"kind\":\"contract\",\"name\":\"" + Sort.escape(name) + "\"");
            sb.append(",\"outBinding\":\"").append(Sort.escape(outBinding)).append("\"");
            if (pre != null) sb.append(",\"pre\":").append(pre.toJson());
            if (post != null) sb.append(",\"post\":").append(post.toJson());
            if (inv != null) sb.append(",\"inv\":").append(inv.toJson());
            if (evidence != null) sb.append(",\"evidence\":\"").append(Sort.escape(evidence)).append("\"");
            sb.append("}");
            return sb.toString();
        }
    }
}

File: implementations/php/provekit-ir-symbolic/src/Ir/Declaration.php
<?php
/** ProvekIt IR — Declaration types and collector pattern. */

namespace ProvekIt\Ir;

class ContractDecl implements \JsonSerializable {
    public function __construct(
        public readonly string $name,
        public readonly string $outBinding = 'out',
        public readonly ?IrFormula $pre = null,
        public readonly ?IrFormula $post = null,
        public readonly ?IrFormula $inv = null,
    ) {}

    public function jsonSerialize(): array {
        $r = ['kind' => 'contract', 'name' => $this->name, 'outBinding' => $this->outBinding];
        if ($this->pre !== null)  $r['pre'] = $this->pre->jsonSerialize();
        if ($this->post !== null) $r['post'] = $this->post->jsonSerialize();
        if ($this->inv !== null)  $r['inv'] = $this->inv->jsonSerialize();
        return $r;
    }
}

class BridgeDecl implements \JsonSerializable {
    public function __construct(
        public readonly string $name,
        public readonly string $sourceSymbol,
        public readonly string $sourceLayer,
        public readonly string $sourceContractCid,
        public readonly string $targetContractCid,
        public readonly string $targetProofCid,
        public readonly string $targetLayer,
        public readonly ?string $notes = null,
    ) {}

    public function jsonSerialize(): array {
        $r = [
            'kind' => 'bridge',
            'name' => $this->name,
            'sourceSymbol' => $this->sourceSymbol,
            'sourceLayer' => $this->sourceLayer,
            'sourceContractCid' => $this->sourceContractCid,
            'targetContractCid' => $this->targetContractCid,
            'targetProofCid' => $this->targetProofCid,
            'targetLayer' => $this->targetLayer,
        ];
        if ($this->notes !== null) $r['notes'] = $this->notes;
        return $r;
    }
}

// ---------- Collector (side-effect-based declaration collection) ----------

class Collector {
    /** @var ContractDecl[] */
    private static array $contracts = [];
    /** @var BridgeDecl[] */
    private static array $bridges = [];

    public static function reset(): void {
        self::$contracts = [];
        self::$bridges = [];
    }

    public static function Contract(
        string $name,
        ?IrFormula $pre = null,
        ?IrFormula $post = null,
        ?IrFormula $inv = null,
        string $outBinding = 'out',
    ): void {
        self::$contracts[] = new ContractDecl($name, $outBinding, $pre, $post, $inv);
    }

    /** Abbreviated: Must(name, formula) -> Contract(name, pre=formula) */
    public static function Must(string $name, IrFormula $pre): void {
        self::$contracts[] = new ContractDecl($name, 'out', $pre, null, null);
    }

    public static function Bridge(
        string $name,
        string $sourceSymbol,
        string $sourceLayer,
        string $sourceContractCid,
        string $targetContractCid,
        string $targetProofCid,
        string $targetLayer,
        ?string $notes = null,
    ): void {
        self::$bridges[] = new BridgeDecl(
            $name, $sourceSymbol, $sourceLayer,
            $sourceContractCid, $targetContractCid,
            $targetProofCid, $targetLayer, $notes
        );
    }

    /** @return array{contracts: ContractDecl[], bridges: BridgeDecl[]} */
    public static function finish(): array {
        $result = ['contracts' => self::$contracts, 'bridges' => self::$bridges];
        self::reset();
        return $result;
    }
}
